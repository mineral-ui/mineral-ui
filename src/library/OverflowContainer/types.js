/* @flow */
import type {
  ComponentTheme,
  ComponentThemeFn,
  ThemeValue
} from '../themes/types';

export type OverflowContainerProps = {
  children?: React$Node,
  containerRef?: (node: HTMLElement) => void,
  hideScrollbars?: boolean,
  scrollX?: boolean,
  scrollY?: boolean,
  tabIndex?: number | string
};

export type OverflowContainerState = {
  scrollable: boolean
};

export type OverflowContainerWithShadowsProps = {
  children?: React$Node,
  hideScrollbars?: boolean,
  scrollX?: boolean,
  scrollY?: boolean
};

export type OverflowContainerWithShadowsState = {
  hasShadowBottom: boolean,
  hasShadowLeft: boolean,
  hasShadowRight: boolean,
  hasShadowTop: boolean
};

export type OverflowContainerThemeFn = ComponentThemeFn<OverflowContainerTheme>;
export type OverflowContainerTheme = ComponentTheme<OverflowContainerThemeKeys>;
type OverflowContainerThemeKeys = {|
  OverflowContainer_outline_focus: ThemeValue
|};

export type OverflowContainerWithShadowsThemeFn = ComponentThemeFn<
  OverflowContainerWithShadowsTheme
>;
export type OverflowContainerWithShadowsTheme = ComponentTheme<
  OverflowContainerWithShadowsThemeKeys
>;
type OverflowContainerWithShadowsThemeKeys = {|
  OverflowContainerWithShadows_outline_focus: ThemeValue,
  OverflowContainerWithShadows_boxShadowBottom: ThemeValue,
  OverflowContainerWithShadows_boxShadowLeft: ThemeValue,
  OverflowContainerWithShadows_boxShadowRight: ThemeValue,
  OverflowContainerWithShadows_boxShadowTop: ThemeValue
|};
