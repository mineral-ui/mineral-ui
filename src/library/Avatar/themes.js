/* @flow */
import type { AvatarThemeFn } from './types';

export const avatarTheme: AvatarThemeFn = (baseTheme) => ({
  Avatar_fontSize_small: baseTheme.fontSize_mouse,
  Avatar_fontSize_medium: baseTheme.fontSize_ui,
  Avatar_fontSize_large: baseTheme.h4_fontSize,
  Avatar_fontSize_jumbo: baseTheme.h4_fontSize,
  Avatar_fontWeight: baseTheme.fontWeight_bold,
  Avatar_size_small: baseTheme.size_small,
  Avatar_size_medium: baseTheme.size_medium,
  Avatar_size_large: baseTheme.size_large,
  Avatar_size_jumbo: baseTheme.size_jumbo,

  ...baseTheme
});
