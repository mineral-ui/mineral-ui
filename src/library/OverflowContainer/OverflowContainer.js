/* @flow */
import React, { Component } from 'react';
import debounce from 'lodash.debounce';
import EventListener from '../EventListener';
import { createStyledComponent } from '../styles';

type Props = {
  children?: React$Node,
  containerRef?: (node: HTMLElement) => void,
  hideScrollbars?: boolean,
  scrollX?: boolean,
  scrollY?: boolean,
  tabIndex?: number | string
};

type State = {
  scrollable: boolean
};

// prettier-ignore
export const componentTheme = (baseTheme: Object) => ({
  OverflowContainer_outline_focus: `1px solid ${baseTheme.borderColor_theme_focus}`,

  ...baseTheme
});

const Root = createStyledComponent(
  'div',
  ({ hideScrollbars, scrollX, scrollY, theme: baseTheme }) => {
    const theme = componentTheme(baseTheme);

    return {
      outline: 0,
      overflowX: scrollX ? 'auto' : undefined,
      overflowY: scrollY ? 'auto' : undefined,
      // Prevent flash of focus style when interacting with children
      transition: 'outline 0.1s 0.25s',

      '&:focus': {
        outline: theme.OverflowContainer_outline_focus
      },

      ...(hideScrollbars
        ? {
            overflow: '-moz-scrollbars-none',
            '-ms-overflow-style': 'none',

            '&::-webkit-scrollbar': {
              display: 'none'
            }
          }
        : undefined)
    };
  },
  {
    displayName: 'OverflowContainer'
  }
);

/**
 * OverflowContainer
 */
export default class OverflowContainer extends Component<Props, State> {
  state = {
    scrollable: false
  };

  container: ?HTMLElement;

  componentDidMount() {
    this.updateScrollable();
  }

  componentDidUpdate() {
    this.updateScrollable();
  }

  setContainerRef = (node: HTMLElement) => {
    const { containerRef } = this.props;

    this.container = node;
    containerRef && containerRef(node);
  };

  updateScrollable = () => {
    const { scrollX, scrollY } = this.props;
    const node = this.container;
    if (node) {
      const scrollable = Boolean(
        (scrollX && node.scrollWidth > node.clientWidth) ||
          (scrollY && node.scrollHeight > node.clientHeight)
      );
      if (this.state.scrollable !== scrollable) {
        this.setState({
          scrollable
        });
      }
    }
  };

  handleWindowResize = debounce(this.updateScrollable, 500);

  render() {
    const { children, tabIndex, ...restProps } = this.props;

    const rootProps = {
      innerRef: this.setContainerRef,
      ...(tabIndex !== undefined
        ? { tabIndex }
        : this.state.scrollable
          ? { tabIndex: 0 }
          : undefined),
      ...restProps
    };
    return (
      <Root {...rootProps}>
        {children}
        <EventListener
          listeners={[
            {
              target: 'window',
              event: 'resize',
              handler: this.handleWindowResize
            }
          ]}
        />
      </Root>
    );
  }
}
