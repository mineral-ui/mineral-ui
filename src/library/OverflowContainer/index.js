/* @flow */
export { default } from './OverflowContainer';
export {
  default as OverflowContainerWithShadows
} from './OverflowContainerWithShadows';
