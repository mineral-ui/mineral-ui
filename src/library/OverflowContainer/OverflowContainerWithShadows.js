/* @flow */
import React, { Component } from 'react';
import debounce from 'lodash.debounce';
import { createStyledComponent } from '../styles';
import { createThemedComponent, mapComponentThemes } from '../themes';
import _OverflowContainer, {
  componentTheme as overflowContainerComponentTheme
} from './OverflowContainer';

type Props = {
  children?: React$Node,
  hideScrollbars?: boolean,
  scrollX?: boolean,
  scrollY?: boolean
};

type State = {
  hasShadowBottom: boolean,
  hasShadowLeft: boolean,
  hasShadowRight: boolean,
  hasShadowTop: boolean
};

// prettier-ignore
export const componentTheme = (baseTheme: Object) =>
  mapComponentThemes(
    {
      name: 'OverflowContainer',
      theme: overflowContainerComponentTheme(baseTheme)
    },
    {
      name: 'OverflowContainerWithShadows',
      theme: {
        OverflowContainerWithShadows_boxShadowBottom: `inset 0 -8px 8px -8px ${baseTheme.color_gray_60}, inset 0 -1px ${baseTheme.borderColor}`,
        OverflowContainerWithShadows_boxShadowLeft: `inset 8px 0 8px -8px ${baseTheme.color_gray_60}, inset 1px 0 ${baseTheme.borderColor}`,
        OverflowContainerWithShadows_boxShadowRight: `inset -8px 0 8px -8px ${baseTheme.color_gray_60}, inset -1px 0 ${baseTheme.borderColor}`,
        OverflowContainerWithShadows_boxShadowTop: `inset 0 8px 8px -8px ${baseTheme.color_gray_60}, inset 0 1px ${baseTheme.borderColor}`,
      }
    },
    baseTheme
  );

const ThemedOverflowContainer = createThemedComponent(
  _OverflowContainer,
  ({ theme: baseTheme }) =>
    mapComponentThemes(
      {
        name: 'OverflowContainerWithShadows',
        theme: componentTheme(baseTheme)
      },
      {
        name: 'OverflowContainer',
        theme: {}
      },
      baseTheme
    )
);

const getBoxShadows = ({ theme: baseTheme, ...restProps }) => {
  const theme = componentTheme(baseTheme);
  const boxShadows = Object.keys(restProps)
    .filter((prop) => prop.indexOf('hasShadow') !== -1)
    .reduce((acc, prop) => {
      if (restProps[prop]) {
        const edge = prop.split('hasShadow')[1];
        acc.push(theme[`OverflowContainerWithShadows_boxShadow${edge}`]);
      }
      return acc;
    }, []);
  return boxShadows.length ? boxShadows.join(',') : undefined;
};

// [1] - 1px to avoid unwanted scrollbar,
//       1px in to avoid potentially cutting off of focus ring of
//       subcomponents in body
const styles = {
  scroller: ({ scrollX, scrollY }) => ({
    flex: '1 1 auto',

    ...(scrollX
      ? {
          overflowY: 'hidden',
          paddingLeft: 2, // [1]
          paddingRight: 2 // [1]
        }
      : undefined),

    ...(scrollY
      ? {
          overflowX: 'hidden',
          paddingBottom: 2, // [1]
          paddingTop: 2 // [1]
        }
      : undefined)
  }),
  root: (props) => {
    const boxShadow = getBoxShadows(props);
    return boxShadow
      ? {
          display: 'flex',
          position: 'relative',

          '&::before': {
            bottom: 0,
            boxShadow,
            content: '""',
            pointerEvents: 'none',
            position: 'absolute',
            left: 0,
            right: 0,
            top: 0
          }
        }
      : undefined;
  }
};

const Root = createStyledComponent('div', styles.root, {
  displayName: 'OverflowContainerWithShadows'
});
const Scroller = createStyledComponent(
  ThemedOverflowContainer,
  styles.scroller,
  {
    displayName: 'Scroller',
    forwardProps: ['containerRef']
  }
);

/**
 * OverflowContainerWithShadows
 */
export default class OverflowContainerWithShadows extends Component<
  Props,
  State
> {
  container: ?HTMLElement;

  setContainerRef = (node: HTMLElement) => {
    this.container = node;
  };

  componentDidMount() {
    this.applyShadows(this.container);
  }

  render() {
    const {
      children,
      hideScrollbars,
      scrollX,
      scrollY,
      ...restProps
    } = this.props;
    const isScrollable = Boolean(
      this.state &&
        Object.keys(this.state).filter((key) => this.state[key]).length
    );
    const rootProps = {
      ...this.state,
      ...restProps
    };
    const scrollerProps = {
      hideScrollbars,
      containerRef: this.setContainerRef,
      onScroll: (scrollX || scrollY) && this.handleScroll,
      scrollX,
      scrollY,
      // Set tabIndex when scrollable so user can scroll with keyboard
      tabIndex: isScrollable ? 0 : undefined
    };

    return (
      <Root {...rootProps}>
        <Scroller {...scrollerProps}>{children}</Scroller>
      </Root>
    );
  }

  handleScroll = (event: SyntheticEvent<HTMLElement>) => {
    this.applyShadows(event.currentTarget);
  };

  applyShadows = debounce(
    (element: HTMLElement) => {
      const { scrollX, scrollY } = this.props;
      if (scrollX) {
        this.setState({
          hasShadowLeft: element.scrollLeft > 0,
          hasShadowRight:
            element.scrollLeft + element.clientWidth < element.scrollWidth
        });
      }
      if (scrollY) {
        this.setState({
          hasShadowTop: element.scrollTop > 0,
          hasShadowBottom:
            element.scrollTop + element.clientHeight < element.scrollHeight
        });
      }
    },
    100,
    { leading: true }
  );
}
