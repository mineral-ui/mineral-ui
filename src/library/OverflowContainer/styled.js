/* @flow */
import styled from '@emotion/styled';
import { overflowContainerTheme } from './themes';
import { overflowContainerWithShadowsTheme } from './themes';

export const OverflowContainerRoot = styled('div')(
  ({ border, borderRadius, hideScrollbars, scrollX, scrollY, theme: baseTheme }) => {
    const theme = overflowContainerTheme(baseTheme);

    return {
      outline: 0,
      overflowX: scrollX ? 'auto' : undefined,
      overflowY: scrollY ? 'auto' : undefined,
      // Prevent flash of focus style when interacting with children
      transition: 'outline 0.1s 0.25s',

      '&:focus': {
        outline: theme.OverflowContainer_outline_focus
      },

      ...(hideScrollbars
        ? {
            overflow: '-moz-scrollbars-none',
            msOverflowStyle: 'none',

            '&::-webkit-scrollbar': {
              display: 'none'
            }
          }
        : undefined),

      /* TargetX Custom Styles */
      ...(border && { border }),
      ...(borderRadius && { borderRadius })
    };
  }
);

export const OverflowContainerWithShadowsRoot = styled('div')((props) => {
  const boxShadow = getBoxShadows(props);
  return boxShadow
    ? {
        display: 'flex',
        position: 'relative',

        '&::before': {
          bottom: 0,
          boxShadow,
          content: '""',
          pointerEvents: 'none',
          position: 'absolute',
          left: 0,
          right: 0,
          top: 0
        }
      }
    : undefined;
});

const getBoxShadows = ({ theme: baseTheme, ...restProps }) => {
  const theme = overflowContainerWithShadowsTheme(baseTheme);
  const boxShadows = Object.keys(restProps)
    .filter((prop) => prop.indexOf('hasShadow') !== -1)
    .reduce((acc, prop) => {
      if (restProps[prop]) {
        const edge = prop.split('hasShadow')[1];
        acc.push(theme[`OverflowContainerWithShadows_boxShadow${edge}`]);
      }
      return acc;
    }, []);
  return boxShadows.length ? boxShadows.join(',') : undefined;
};

export const getScrollerStyles = ({
  scrollX,
  scrollY
}: {
  scrollX: boolean,
  scrollY: boolean
}) => ({
  flex: '1 1 auto',

  // [1] - 1px to avoid unwanted scrollbar,
  //       1px in to avoid potentially cutting off of focus ring of
  //       subcomponents in body
  ...(scrollX
    ? {
        overflowY: 'hidden',
        paddingLeft: 2, // [1]
        paddingRight: 2 // [1]
      }
    : undefined),

  ...(scrollY
    ? {
        overflowX: 'hidden',
        paddingBottom: 2, // [1]
        paddingTop: 2 // [1]
      }
    : undefined)
});
