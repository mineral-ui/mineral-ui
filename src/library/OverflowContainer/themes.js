/* @flow */
import { mapComponentThemes } from '../themes';

import type {
  OverflowContainerThemeFn,
  OverflowContainerWithShadowsThemeFn
} from './types';

// prettier-ignore
export const overflowContainerTheme: OverflowContainerThemeFn = (baseTheme) => ({
  OverflowContainer_outline_focus: `1px solid ${baseTheme.borderColor_theme_focus}`,

  ...baseTheme
});

// prettier-ignore
// $FlowFixMe - strict theme keys incompatible with mapComponentThemes
export const overflowContainerWithShadowsTheme: OverflowContainerWithShadowsThemeFn = (baseTheme) =>
  mapComponentThemes(
    {
      name: 'OverflowContainer',
      theme: overflowContainerTheme(baseTheme)
    },
    {
      name: 'OverflowContainerWithShadows',
      theme: {
        OverflowContainerWithShadows_boxShadowBottom: `inset 0 -8px 8px -8px ${baseTheme.color_gray_60}, inset 0 -1px ${baseTheme.borderColor}`,
        OverflowContainerWithShadows_boxShadowLeft: `inset 8px 0 8px -8px ${baseTheme.color_gray_60}, inset 1px 0 ${baseTheme.borderColor}`,
        OverflowContainerWithShadows_boxShadowRight: `inset -8px 0 8px -8px ${baseTheme.color_gray_60}, inset -1px 0 ${baseTheme.borderColor}`,
        OverflowContainerWithShadows_boxShadowTop: `inset 0 8px 8px -8px ${baseTheme.color_gray_60}, inset 0 1px ${baseTheme.borderColor}`,
      }
    },
    baseTheme
  );
