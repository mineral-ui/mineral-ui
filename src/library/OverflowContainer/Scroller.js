/* @flow */
import styled from '@emotion/styled';
import { mapComponentThemes, themed } from '../themes';
import OverflowContainer from './OverflowContainer';
import { getScrollerStyles } from './styled';
import { overflowContainerWithShadowsTheme } from './themes';

import type { StyledComponent } from '@emotion/styled-base/src/utils';

// NOTE: These components cannot live in styled.js due to circular dependency
// issues with OverflowContainer

const ThemedOverflowContainer = themed(OverflowContainer)(
  ({ theme: baseTheme }) =>
    mapComponentThemes(
      {
        name: 'OverflowContainerWithShadows',
        theme: overflowContainerWithShadowsTheme(baseTheme)
      },
      {
        name: 'OverflowContainer',
        theme: {}
      },
      baseTheme
    )
);

const Scroller: StyledComponent<{ [key: string]: any }> = styled(
  ThemedOverflowContainer
)(getScrollerStyles);

export default Scroller;
