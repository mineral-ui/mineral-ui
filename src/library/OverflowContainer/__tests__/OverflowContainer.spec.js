/* @flow */
import React from 'react';
import { shallow } from 'enzyme';
import { mineralTheme } from '../../themes';
import { getProcessedComponentThemeKeys } from '../../themes/processComponentTheme';
import OverflowContainer, { componentTheme } from '../OverflowContainer';
import { mountInThemeProvider } from '../../../../utils/enzymeUtils';
import testThemeOverrides from '../../../../utils/testThemeOverrides';

/*
 * [1] This getter does not exist on the HTMLElement.prototype in JSDOM, so we
 *     must mock it on the global. `clientWidth`, also used by OverflowContainer
 *     always returns 0, which is conveniently useful.
 * [2] Then we must clean up our global mocking in [1]
 */

const defaultProps = {
  scrollX: true
};

const shallowOverflowContainer = (props = {}) => {
  const containerProps = {
    ...defaultProps,
    ...props
  };
  return shallow(<OverflowContainer {...containerProps} />);
};

describe('OverflowContainer', () => {
  it('renders', () => {
    const overflowContainer = shallowOverflowContainer();

    expect(overflowContainer.exists()).toEqual(true);
  });

  describe('theme overrides', () => {
    testThemeOverrides(
      <OverflowContainer />,
      getProcessedComponentThemeKeys(componentTheme(mineralTheme))
    );
  });

  it('applies tabIndex=0 when scrollable', () => {
    // [1]
    // $FlowFixMe
    Object.defineProperties(window.HTMLElement.prototype, {
      scrollWidth: {
        // Note that this mocked getter is designed to always trigger overflow
        // because it's greater than offsetWidth
        get: () => 10,
        configurable: true
      }
    });

    const [, overflowContainer] = mountInThemeProvider(
      <OverflowContainer {...defaultProps} />
    );

    // [2]
    // $FlowFixMe
    Object.defineProperties(window.HTMLElement.prototype, {
      scrollWidth: {
        get: undefined
      }
    });

    expect(overflowContainer.html()).toMatchSnapshot();
  });
});
