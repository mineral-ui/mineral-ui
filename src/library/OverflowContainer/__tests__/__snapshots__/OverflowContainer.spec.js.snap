// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowContainer applies tabIndex=0 when scrollable 1`] = `"<div class=\\"css-oq2iuk-OverflowContainer\\" tabindex=\\"0\\"></div>"`;
